specs/DisableNavigatingBackAndForward.md

```diff
@@ -0,0 +1,129 @@
+# API spec for disable navigating back/forward
+
+# Background
+This problem was first proposed by a developer on GitHub, who wants to prevent users navigating
+back or forward using any of the built-in shortcut keys or special mouse buttons.
+
+Afterwards, Teams made similar demands. They wanted a mechanism which could support them in
+controlling the behaviors of `go back` and `go forward` freely, like disabling them.
+
+@Haichao Zhu has already finished some work on letting application developers handle all input and
+decide whether to suppress. 
+
+This should be solvable in a generic way. However, this feature hasn’t been released yet, and it might
+be better if we could provide a simpler and more direct way. 
+
+Therefore, our job is to provide a mechanism for developers to disable navigating back and forward
+without excessive effort.
+
+
+# Examples
+#### Win32 C++
+
+##### Use `ICoreWebView2NavigationStartingEventArgs3` in `add_NavigationStarting`
+
+```c++
+//! [NavigationStarting]
+// Register a handler for the NavigationStarting event.
+// This handler will check the navigation status, and if the navigation is
+// `GoBack` or `GoForward`, it will be canceled.
+CHECK_FAILURE(m_webView->add_NavigationStarting(
+    Callback<ICoreWebView2NavigationStartingEventHandler>(
+        [this](ICoreWebView2* sender, ICoreWebView2NavigationStartingEventArgs* args)
+        -> HRESULT {
+            wil::com_ptr<ICoreWebView2NavigationStartingEventArgs3> args3;
+            if (SUCCEEDED(args->QueryInterface(IID_PPV_ARGS(&args3))))
+            {
+                COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND history_change =
+                    COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND_OTHER;
+                CHECK_FAILURE(args3->get_NavigationHistoryChange(&history_change));
+                if (history_change != COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND_OTHER)
```

Comment:
I think "Programmatic" is a better name than "Other".

"Other" suggests that this is a catch-all for navigations we don't understand. But really, we do understand it. It was a programmatic navigation.

It also avoids confusion if we later add another kind of navigation, say "Inversion", and now we have two kinds of "Other" to report to downlevel clients. We have "Old Other" (programmatic navigation) and "New Other" (Inversion navigation or whatever).

I think we should reserve "Other" to mean "Things for which we cannot assign a category."

Also, do Refreshes show up as a navigation?

Reply:
If people care to distinguish, then maybe two Boolean properties `IsForwardNavigation` and `IsBackwardNavigation`.

If people don't care about the direction and just care about the fact that it was a history navigation then maybe `IsHistoryNavigation`? (Assuming this would be `true` for `history.go(-3)` too.)

And we can decide whether a reload counts as a history navigation. (I think the answer is "No".)

Reply:
I'm having trouble understanding what exactly what's wanted. It sounds like we want to be able to cancel, regardless what the trigger, moving through the history. But navigation to a _new_ location, as opposed to navigating forward in the history stack, is OK?

Reply:
Yes that's pretty much it. Teams (and one other customer) want to stop the end user from being able to navigate the webview back/forward using the built-in accelerator keys. Those apps also don't do back/forward programmatically so they can just check if this is back/forward and if so cancel the navigation. If they did back/forward programmatically then they could additionally check the user IsUserInitiated property and cancel only user initiated back/forward navigations.

Reply:
> And we can decide whether a reload counts as a history navigation. (I think the answer is "No".)

I agree.

> If people care to distinguish, then maybe two Boolean properties IsForwardNavigation and IsBackwardNavigation.
> If people don't care about the direction and just care about the fact that it was a history navigation then maybe IsHistoryNavigation? (Assuming this would be true for history. Go(-3) too.)

Both of those seem good to me. These customers don't require the extra information of back or forward in this case, but it doesn't hurt to provide this extra info with the two IsForward/BackwardNavigation Boolean properties. It also aligns better with how we express back/forward already: CoreWebView2.CanGoBack/CanGoForward and CoreWebView2.GoBack()/GoForward()

Interested to hear other's thoughts of course, but otherwise I like changing this to Raymond's first suggestion of 
```
Boolean IsForwardNavigation { get; };
Boolean IsBackwardNavigation { get; };
```

Reply:
I like those

Reply:
IsBackNavigation to align with GoBack()/CanGoBack? Or is that too grammatically awkward?

Reply:
I prefer using the `enum`, which could reveal more information and be easy to expand.
I need test whether we could distinguish `{ back, forward, reload, "new" navigation }`.

Reply:
I would choose int HistoryNavigationOffset.

Reply:
I like using Back instead of Backward to match existing APIs, but not strongly.

@plantree I don't think we know the full set of things that should go in that enum. Or if its practical to implement. If you know now if its practical to implement using the enum described please let us know. I've found this [DOM API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming/type) which is similar and has the same types except the prefetch hint type which isn't applicable to us. This could be like `enum NavigationMethodKind { Back, Forward, Reload, Navigate }` where Reload would mean a reload kind of method like `CoreWebView2.Reload`, or `window.location.reload`, or the user invoking reload via input.

Otherwise the two Booleans work well

```
Boolean IsForwardNavigation { get; };
Boolean IsBackNavigation { get; };
```

Reply:
`HistoryNavigationOffset` doesn't seem to be enough. For example, if you open a new url, the `offset` will be `1`. Perhaps we need combine the state of commit.

Reply:
Hi @david-risney,
I have two new updates:
1. It seems that `IsUserInitiated` cannot accurately reveal that navigation is triggered by the user.
2. `GetNavigationEntryOffset()` will be confusing between `forward` and `new entry`, which will similarly `return 1`. However, I find one alternative implementation, which is [navigation_params.mojom](https://chromium.googlesource.com/chromium/src/+/def301837d32598a933808e3e819c909be7706d3/content/common/navigation_params.mojom). We could use `NavigationType` to distinguish `back/forward` and other types.

Reply:
Let me follow up with you offline and then we'll get back to the API review after.

Reply:
Hi @david-risney,
I have pushed a new commit, which use `NavigationKind` enum to classify different navigations.
`NavigationType` in `navigation_params.mojom` is a very userful source, and this new design will be added to the design doc later.